lib/src/dart2_9_suggestors/dart2_9_utilities.dart

```diff
@@ -0,0 +1,170 @@
+///   // ...
+/// )(_$Foo); // [argList]
+/// ```
+///
+/// Returns `null` if no generated argument is found or if [argList] is not used
+/// with `connect`.
+///
+/// Example:
+///
+/// | input                          | output |
+/// |--------------------------------|--------|
+/// | (_$Foo)                        | _$Foo  |
+/// | (castUiFactory(_$Foo))         | _$Foo  |
+/// | (Foo)                          | null   |
+SimpleIdentifier getConnectGeneratedFactoryArg(ArgumentList argList) {
+  final args = argList.arguments;
+  if (args.length != 1) return null;
+
+  final generatedPrefix = '_\$';
+
+  final method =
+      argList.thisOrAncestorOfType<FunctionExpressionInvocation>()?.function;
+  if (method is MethodInvocation && method.methodName.name == 'connect') {
+    final generatedArg = args.first;
+    if (generatedArg is SimpleIdentifier) {
+      return generatedArg.name.startsWith(generatedPrefix)
+          ? generatedArg
+          : null;
+    } else if (generatedArg is MethodInvocation &&
+        generatedArg.methodName.name == castFunctionName) {
+      final arg = generatedArg.argumentList.arguments.first;
+      return arg is SimpleIdentifier && arg.name.startsWith(generatedPrefix)
+          ? arg
+          : null;
+    }
+  }
+
+  return null;
+}
+
+/// Returns the generated factory config argument from [argList].
+///
+/// Used specifically to find the generated factory config in `uiFunction` /
+/// `uiForwardRef` usages.
+///
+/// Returns `null` if no generated argument found.
+///
+/// Example:
+///
+/// | input                                                  | output      |
+/// |--------------------------------------------------------|-------------|
+/// | ((props) {}, $FooConfig)                               | $FooConfig  |
+/// | ((props) {}, _$FooConfig)                              | _$FooConfig |
+/// | ((props) {}, UiFactoryConfig())                        | null        |
+SimpleIdentifier getGeneratedFactoryConfigArg(ArgumentList argList) {
+  final args = argList.arguments;
+  if (args.length != 2) return null;
+
+  final configPattern = RegExp(r'_?\$\S*Config$');
+
+  final generatedArg = args[1];
+  return generatedArg is SimpleIdentifier &&
+          generatedArg.name.startsWith(configPattern)
+      ? generatedArg
+      : null;
+}
+
+/// Returns whether or not [node] is a class component factory declaration.
+bool isClassComponentFactory(TopLevelVariableDeclaration node) {
+  final type = node.variables?.type;
+  if (type != null && type is NamedType && type?.name?.name == 'UiFactory') {
+    final initializer = node.variables?.variables?.first?.initializer;
+    if (initializer is SimpleIdentifier) {
+      return initializer.name.startsWith('_\$');
+    } else if (initializer is MethodInvocation &&
+        initializer.methodName.name == castFunctionName) {
+      return initializer.argumentList.arguments.first
+          .toSource()
+          .startsWith('_\$');
+    }
```

Comment:
This could be simplified, and cover more cases (such as `connect` and wrapper functions, as well as deeply nested wrapper functions), by checking to see if the generated factory is used anywhere within the initializer (which is the same as what we do in the builder):
```suggestion
    if (initializer != null) {
      final name = node.variables.variables.first.name.name;
      final generatedName = r'_$' + name;
      return allDescendantsOfType<SimpleIdentifier>(initializer)
          .any((identifier) => identifier.name == generatedName);
    }
```
(using allDescendantsOfType [from over_react](https://github.com/Workiva/over_react/blob/ac85e9d2fdbd7fa7f9f0174c845e3c8951b46e95/tools/analyzer_plugin/lib/src/util/ast_util.dart#L40-L57))

A similar approach could be taken to find the identifier where `getConnectGeneratedFactoryArg` is used, allowing you to cover both the basic and `connect` cases with the same code.

I think this improvement would be worth it, but feel free to disagree.

Reply:
That does look nicer! Right now over_react_codemod doesn't have over_react as a dependency - is it worth adding the dependency for one utility? 

Also is `allDescendantsOfType` exported anywhere in over_react? I can't seem to access it, but I might be missing something.

Reply:
Oh sorry, I should have clarified; I'd recommend just copy-pasting the code from that repo, since it's such a small implementation.

Reply:
Ohh 👍